matrix/nth_fibonnacci_using_matrix_exponentiation.py

```diff
@@ -0,0 +1,100 @@
+"""
+Implementation of finding nth fibonacci number using matrix exponentiation.
+Time Complexity is about O(log(n)*8)
+As we know
+    f[n] = f[n-1] + f[n-1]
+Converting to matrix,
+    [f(n),f(n-1)] = [[1,1],[1,0]] * [f(n-1),f(n-2)]
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^2 * [f(n-2),f(n-3)]
+    ...
+    ...
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^(n-1) * [f(1),f(0)]
+
+So we just need the n times multiplication of  the matrix [1,1],[1,0]].
+We can decrease the n times multiplication by following  the divide and conquer approach
+
+"""
+from __future__ import print_function
+
+
+def multiply(matrix_a, matrix_b):
+    matrix_c = []
+    n = len(matrix_a)
+    for i in range(n):
+        list_1 = []
+        for j in range(n):
+            val = 0
+            for k in range(n):
+                val = val + matrix_a[i][k] * matrix_b[k][j]
+            list_1.append(val)
+        matrix_c.append(list_1)
+    return matrix_c
+
+
+def identity(n):
+    return [[int(row == column) for column in range(n)] for row in range(n)]
+
+
+def zerro(n):
+    return [[int(row == column) for column in range(n)] for row in range(n)]
+
+
+def nth_fibonacci(n):
+    if n <= 1:
+        return n
+    res_matrix = identity(2)
+    fibonacci_matrix = [[1, 1], [1, 0]]
+    n = n - 1
+    while n > 0:
+        if n % 2 == 1:
+            res_matrix = multiply(res_matrix, fibonacci_matrix)
+        fibonacci_matrix = multiply(fibonacci_matrix, fibonacci_matrix)
+        n = int(n / 2)
+    return res_matrix[0][0]
+
+
+def nth_fibonnaci_test(n):
+    if n <= 1:
+        return n
+    fib0 = 0
+    fib1 = 1
+    for i in range(2, n + 1):
+        tem = fib1 + fib0
+        fib0 = fib1
+        fib1 = tem
```

Comment:
In Python, these three lines could be replaced one line with:
```python
fib0, fib1 = fib1, fib0 + fib1
```
The right hand side will get fully evaluated before the variable assignment happens on the left hand side.

Reply:
Resolved . Thanks  a lot .